pkg/kube/inject/inject.go

```diff
@@ -278,6 +289,99 @@ func injectRequired(ignored []string, config *Config, podSpec *corev1.PodSpec, m
 	return required
 }
 
+// ProxyImage constructs image url in a backwards compatible way.
+// values based name => {{ .Values.global.hub }}/{{ .Values.global.proxy.image }}:{{ .Values.global.tag }}
+func ProxyImage(values *opconfig.Values, image *proxyConfig.ProxyImage, annotations map[string]string) string {
+	imageName := "proxyv2"
+	global := &opconfig.GlobalConfig{}
+	if values != nil && values.Global != nil {
+		global = values.Global
+	}
+
+	hub := global.Hub
+	tag := ""
+	if stag, ok := global.Tag.(string); ok {
+		tag = stag
+	}
+
+	imageType := ""
+
+	if image != nil {
+		imageType = image.ImageType
+	}
+
+	if it, ok := annotations[annotation.SidecarProxyImageType.Name]; ok {
+		imageType = it
+	}
+
+	return imageURL(hub, imageName, tag, imageType)
+}
+
+// parseTag returns components of the tag
+// tag := {istio_tag}-{vendor_tag}-{imageType}
+// examples:
+// 1.11.2-debug
+// 1.11.2-asm.17-distroless
+func parseTag(tag string) (istioTag string, vendorTag string, imageType string, unusual bool) {
+	tagParts := strings.SplitN(tag, "-", 3)
+	istioTag = tagParts[0]
+	if len(tagParts) == 2 {
+		if strings.Contains(tagParts[1], ".") {
```

Comment:
This part is a bit iffy.
It will work in most cases. The mode of failure is same whether we try to handle this or not.

Reply:
If we use distroless and debug as the allow list, then it works. At that point image type is effectively an enum. I could add a meshConfig.image_types where a customer can add their own stuff, or just say that these are the only two that would be processed or replaced.

If the mesh admin changes the tag to include image type, say ‘-centos’, how do we want the pod to behave? 

I will fix the code to capture “last component” regardless.